From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: csagan5 <32685696+csagan5@users.noreply.github.com>
Date: Sat, 26 Sep 2020 14:23:19 +0100
Subject: DoH improvements

Enable secure mode by default
Always enforce DoH even with inconsistent system DNS configuration

License: GPL-3.0-only - https://spdx.org/licenses/GPL-3.0-only.html
---
 chrome/browser/net/secure_dns_config.cc                  | 2 +-
 chrome/browser/net/secure_dns_util.cc                    | 2 +-
 chrome/browser/net/stub_resolver_config_reader.cc        | 7 +++++++
 chrome/browser/privacy/secure_dns_bridge.cc              | 2 +-
 .../ui/webui/settings/settings_secure_dns_handler.cc     | 2 +-
 net/dns/dns_client.cc                                    | 4 +++-
 net/dns/host_resolver_manager.cc                         | 1 +
 net/dns/public/doh_provider_entry.cc                     | 9 +++------
 net/dns/public/doh_provider_entry.h                      | 4 ++--
 9 files changed, 20 insertions(+), 13 deletions(-)

diff --git a/chrome/browser/net/secure_dns_config.cc b/chrome/browser/net/secure_dns_config.cc
index bfed26a8e2437833ad77549440f9373e2dc0975e..8431257f871e6d85f8d5d4e727e480bb57e0bf4d 100644
--- a/chrome/browser/net/secure_dns_config.cc
+++ b/chrome/browser/net/secure_dns_config.cc
@@ -32,7 +32,7 @@ std::optional<net::SecureDnsMode> SecureDnsConfig::ParseMode(
   if (name == kModeSecure) {
     return net::SecureDnsMode::kSecure;
   } else if (name == kModeAutomatic) {
-    return net::SecureDnsMode::kAutomatic;
+    return net::SecureDnsMode::kSecure;
   } else if (name == kModeOff) {
     return net::SecureDnsMode::kOff;
   }
diff --git a/chrome/browser/net/secure_dns_util.cc b/chrome/browser/net/secure_dns_util.cc
index 453de78e45aceaf33f448e18cc012a1eed5f0496..ab744d202d59a40a21dddcce8a47eca9497588be 100644
--- a/chrome/browser/net/secure_dns_util.cc
+++ b/chrome/browser/net/secure_dns_util.cc
@@ -35,7 +35,7 @@ bool EntryIsForCountry(const net::DohProviderEntry* entry,
         return country_codes::CountryId(country_code) == country_id;
       });
   if (matches) {
-    DCHECK(!entry->ui_name.empty());
+    DCHECK(!entry->ui_name_cromite.empty());
     DCHECK(!entry->privacy_policy.empty());
   }
   return matches;
diff --git a/chrome/browser/net/stub_resolver_config_reader.cc b/chrome/browser/net/stub_resolver_config_reader.cc
index e30c6edaa5f657ef713f248d69232a1712042349..4a0dfdf9f9754ba11aaa70c6c6554a3bf9f08a34 100644
--- a/chrome/browser/net/stub_resolver_config_reader.cc
+++ b/chrome/browser/net/stub_resolver_config_reader.cc
@@ -57,6 +57,7 @@ namespace {
 
 #if BUILDFLAG(IS_WIN)
 bool ShouldDisableDohForWindowsParentalControls() {
+  if ((true)) return false;
   return GetWinParentalControls().web_filter;
 }
 
@@ -243,6 +244,7 @@ void StubResolverConfigReader::UpdateNetworkService(bool record_metrics) {
 }
 
 bool StubResolverConfigReader::ShouldDisableDohForManaged() {
+  if ((true)) return false;
 // This function ignores cloud policies which are loaded on a per-profile basis.
 #if BUILDFLAG(IS_ANDROID)
   // Check for MDM/management/owner apps. android_has_owner_ is true if either a
@@ -271,6 +273,7 @@ bool StubResolverConfigReader::ShouldDisableDohForManaged() {
 }
 
 bool StubResolverConfigReader::ShouldDisableDohForParentalControls() {
+  if ((true)) return false;
   if (parental_controls_testing_override_.has_value())
     return parental_controls_testing_override_.value();
 
@@ -428,6 +431,10 @@ SecureDnsConfig StubResolverConfigReader::GetAndUpdateConfiguration(
   if (secure_dns_mode != net::SecureDnsMode::kOff) {
     doh_config = net::DnsOverHttpsConfig::FromStringLax(
         GetDnsOverHttpsConfigSource()->GetDnsOverHttpsTemplates());
+    if (secure_dns_mode == net::SecureDnsMode::kSecure &&
+          GetDnsOverHttpsConfigSource()->GetDnsOverHttpsTemplates().empty()) {
+      doh_config = net::DnsOverHttpsConfig::FromStringLax("https://doh.opendns.com/dns-query{?dns}");
+    }
     if (ShouldUseDohFallback(secure_dns_mode,
                              *GetDnsOverHttpsConfigSource())) {
       mode_details =
diff --git a/chrome/browser/privacy/secure_dns_bridge.cc b/chrome/browser/privacy/secure_dns_bridge.cc
index 5a47088fb07d3f2085af21a907e88d4743ea5c94..42789586fa22cf497161096865969a4b3190f8a1 100644
--- a/chrome/browser/privacy/secure_dns_bridge.cc
+++ b/chrome/browser/privacy/secure_dns_bridge.cc
@@ -109,7 +109,7 @@ static ScopedJavaLocalRef<jobjectArray> JNI_SecureDnsBridge_GetProviders(
       providers, std::back_inserter(ret),
       [](const net::DohProviderEntry* entry) -> std::vector<std::u16string> {
         net::DnsOverHttpsConfig config({entry->doh_server_config});
-        return {base::UTF8ToUTF16(entry->ui_name),
+        return {base::UTF8ToUTF16(entry->ui_name_cromite.empty() ? entry->provider : entry->ui_name_cromite),
                 base::UTF8ToUTF16(config.ToString()),
                 base::UTF8ToUTF16(entry->privacy_policy)};
       });
diff --git a/chrome/browser/ui/webui/settings/settings_secure_dns_handler.cc b/chrome/browser/ui/webui/settings/settings_secure_dns_handler.cc
index 5b0cad4c754ce751a1d660f6ffc1bd1f2ab128d6..fc100a57bce05bdf7fa20aa667f7c6b2a9a3c6ea 100644
--- a/chrome/browser/ui/webui/settings/settings_secure_dns_handler.cc
+++ b/chrome/browser/ui/webui/settings/settings_secure_dns_handler.cc
@@ -127,7 +127,7 @@ base::ListValue SecureDnsHandler::GetSecureDnsResolverList() {
   for (const net::DohProviderEntry* entry : providers_) {
     net::DnsOverHttpsConfig doh_config({entry->doh_server_config});
     base::DictValue dict;
-    dict.Set("name", entry->ui_name);
+    dict.Set("name", entry->ui_name_cromite.empty() ? entry->provider : entry->ui_name_cromite);
     dict.Set("value", doh_config.ToString());
     dict.Set("policy", entry->privacy_policy);
     resolvers.Append(std::move(dict));
diff --git a/net/dns/dns_client.cc b/net/dns/dns_client.cc
index f54e881686e8f4b340e7a502a59cde1c141434df..2f911f2aa2251414d3d3225c510d1cde6d747b35 100644
--- a/net/dns/dns_client.cc
+++ b/net/dns/dns_client.cc
@@ -329,7 +329,9 @@ class DnsClientImpl : public DnsClient {
  private:
   DnsConfig BuildEffectiveConfig() const {
     DnsConfig config;
-    if (config_overrides_.OverridesEverything()) {
+    // in Bromite it is sufficient to have secure DoH enabled to give the overrides priority
+    if (config_overrides_.dns_over_https_config &&
+        config_overrides_.secure_dns_mode) {
       config = config_overrides_.ApplyOverrides(DnsConfig());
     } else {
       config = config_overrides_.ApplyOverrides(
diff --git a/net/dns/host_resolver_manager.cc b/net/dns/host_resolver_manager.cc
index a1991330f9b2c6413f26053272ae4ffcdc670c3b..eb7a0d675883f83e0262efc44f0c4a5efd720adc 100644
--- a/net/dns/host_resolver_manager.cc
+++ b/net/dns/host_resolver_manager.cc
@@ -689,6 +689,7 @@ void HostResolverManager::SetDnsConfigOverrides(DnsConfigOverrides overrides) {
   bool changed = dns_client_->SetConfigOverrides(std::move(overrides));
 
   if (changed) {
+    LOG(INFO) << "triggering non-system DNS change";
     NetworkChangeNotifier::TriggerNonSystemDnsChange();
 
     // Only invalidate cache if new overrides have resulted in a config change.
diff --git a/net/dns/public/doh_provider_entry.cc b/net/dns/public/doh_provider_entry.cc
index f48d26044b3c9403118930eaee6ae2835dce66b2..a2a874d6b9a4db99b75364893c3eef69737ccaa1 100644
--- a/net/dns/public/doh_provider_entry.cc
+++ b/net/dns/public/doh_provider_entry.cc
@@ -437,15 +437,12 @@ DohProviderEntry::DohProviderEntry(
       doh_server_config(
           ParseValidDohTemplate(std::move(dns_over_https_template),
                                 dns_over_https_server_ip_strs)),
-      ui_name(ui_name),
+      ui_name_cromite(ui_name),
       privacy_policy(privacy_policy),
-      display_globally(display_globally),
+      display_globally(true),
       display_countries(std::move(display_countries)) {
   DCHECK(!display_globally || this->display_countries.empty());
-  if (display_globally || !this->display_countries.empty()) {
-    DCHECK(!this->ui_name.empty());
-    DCHECK(!this->privacy_policy.empty());
-  }
+  if (this->privacy_policy.empty()) this->display_globally = false;
   for (const auto& display_country : this->display_countries) {
     DCHECK_EQ(2u, display_country.size());
   }
diff --git a/net/dns/public/doh_provider_entry.h b/net/dns/public/doh_provider_entry.h
index 93a30451be5a3a7c10c1066d92d2b0fc2ce1f250..2e5c31297a657d1cec61e0cf17929675cabc2f89 100644
--- a/net/dns/public/doh_provider_entry.h
+++ b/net/dns/public/doh_provider_entry.h
@@ -50,7 +50,7 @@ struct NET_EXPORT DohProviderEntry {
   std::set<IPAddress> ip_addresses;
   base::flat_set<std::string_view> dns_over_tls_hostnames;
   DnsOverHttpsServerConfig doh_server_config;
-  std::string_view ui_name;
+  std::string_view ui_name_cromite;
   std::string_view privacy_policy;
   bool display_globally;
   base::flat_set<std::string> display_countries;
@@ -90,7 +90,7 @@ struct NET_EXPORT DohProviderEntry {
       std::initializer_list<std::string_view> dns_over_53_server_ip_strs,
       base::flat_set<std::string_view> dns_over_tls_hostnames,
       std::string dns_over_https_template,
-      std::string_view ui_name,
+      std::string_view ui_name_cromite,
       std::string_view privacy_policy,
       bool display_globally,
       base::flat_set<std::string> display_countries,
